src/Model/ModelManagerInterface.php

```diff
@@ -97,135 +80,102 @@ public function createQuery($class, $alias = 'o');
 
     /**
      * Get the identifiers as a string that is safe to use in a url.
      *
      * This is similar to getNormalizedIdentifier but guarantees an id that can
      * be used in a URL.
-     *
-     * @param object $model
-     *
-     * @return string string representation of the id that is safe to use in a url
      */
-    public function getUrlSafeIdentifier($model);
+    public function getUrlSafeIdentifier(object $model): string;
 
     /**
      * Create a new instance of the model of the specified class.
-     *
-     * @param string $class
-     *
-     * @return object
      */
-    public function getModelInstance($class);
+    public function getModelInstance(string $class): object;
 
     /**
-     * @param string $class
-     *
-     * @return array|\ArrayAccess
+     * @return Collection<array-key, object>
      */
-    public function getModelCollectionInstance($class);
+    public function getModelCollectionInstance(string $class): Collection;
 
     /**
      * Removes an element from the collection.
      *
-     * @param array  $collection
-     * @param object $element
+     * @param Collection<array-key, object> $collection
      */
-    public function collectionRemoveElement(&$collection, &$element);
+    public function collectionRemoveElement(Collection &$collection, object &$element): void;
 
     /**
      * Add an element from the collection.
      *
-     * @param array  $collection
-     * @param object $element
+     * @param Collection<array-key, object> $collection
      */
-    public function collectionAddElement(&$collection, &$element);
+    public function collectionAddElement(Collection &$collection, object &$element): void;
 
     /**
      * Check if the element exists in the collection.
      *
-     * @param array  $collection
-     * @param object $element
-     *
-     * @return bool
+     * @param Collection<array-key, object> $collection
      */
-    public function collectionHasElement(&$collection, &$element);
+    public function collectionHasElement(Collection &$collection, object &$element): bool;
 
     /**
      * Clear the collection.
      *
-     * @param array $collection
+     * @param Collection<array-key, object> $collection
      */
-    public function collectionClear(&$collection);
+    public function collectionClear(Collection &$collection): void;
 
     /**
-     * @param string $class
-     *
-     * @return object
+     * @param array<string, mixed> $array
      */
-    public function modelReverseTransform($class, array $array = []);
+    public function modelReverseTransform(string $class, array $array = []): object;
 
-    /**
-     * @param string $class
-     * @param object $instance
-     *
-     * @return object
-     */
-    public function modelTransform($class, $instance);
+    public function modelTransform(string $class, object $instance): object;
 
     /**
-     * @param mixed $query
+     * @param ProxyQueryInterface|object $query
```

Comment:
I feel like we should remove this method, in favor of the execute method in ProxyQueryInterface

Because, IMHO, all the query in our code should be ProxyQueryInterface.

Reply:
Remove this method could be done with a deprecation first on 3.x and a migration path. I added two type hints because check how it is implemented on Sonata doctrine orm

Reply:
Indeed.

The only use in SonataAdmin is https://github.com/sonata-project/SonataAdminBundle/blob/3.x/src/Form/ChoiceList/ModelChoiceLoader.php#L97

This would means we need to restrict `ModelChoiceLoader` construct to only accept ProxyqueryInterface for the `query` argument. I'm not sure it's the best idea, since users will certainly prefer to use a QueryBuilder.

But I think we could add a method `supports` to the ModelManagerInterface. The implementation would be
```
function supports($query) {
    return $query instanceof ProxyQueryInterface || $query instanceof QueryBuilder;
}
```

Then in the construct of the ModelChoiceLoader, we could add a check
```
if (!$this->modelManager->supports($query)) {
     throw \InvalidArgument(...)
}
```

And we could also chance the implementation of executeQuery to
```
function executeQuery($query) {
if (!$this->supports($query)) {
     throw \InvalidArgument(...)
}

// old implementation
}
```

WDYT ?

Reply:
Like the idea, but typhing here with ProxyQueryInterface|QueryBuilder will force us to declare more dependency with the orm right? What namespace is the QueryBuilder? It is the same for the sonatadoctrinemongodb bundle?

Reply:
There wont need for a QueryBuilder typehint in SonataAdmin.

ModelManagerInterface will have:
```
function supports(mixed $query): boolean
```

ModelChoiceLoader will have the __construct
```
public function __construct(ModelManagerInterface $modelManager, ..., mixed $query, ...) {
    if (!$modelManager->supports($query)) {
         throw \InvalidArgument(...)
    }
}
```

And DoctrineORM will have the implementation 
```
function supports($query) {
    return $query instanceof ProxyQueryInterface || $query instanceof QueryBuilder;
}
```

But DoctrineODM will have the implementation
```
function supports($query) {
    return $query instanceof ProxyQueryInterface || $query instanceof Builder;
}
```

Reply:
Do you agree on leaving this typehint as is or should I change it to mixed? (IMO it is better to know it will be a object anyway, and it could be a `ProxyQueryInterface`

Reply:
I would change this to `object`.

`ProxyQueryInterface|object = object`

Reply:
Changed, not sure if that is what you had in mind, please resolve conversation if it is :)

Reply:
It is